### PR TITLE
Coin dist adventures

### PR DESCRIPTION
**Pull request by darosior**
New algorithms for:
- The coin layout
- The Cancel coin selection
- The CF coin selection

They have a stronger rationale IMO and also much better results in my initial testings. Still WIP though!

### CODE REVIEW

**Comment by darosior** [comment]
We should really ditch all the other coin selection but the one not consolidating.. All the heuristics they use are flawed.

**Comment by darosior** [comment]
Ok, pused an update. I think it's now final although i'm not 100% sure i recall my TODOs from 2 weeks ago..

**Comment by darosior** [comment]
Rebased on master after #93 merge. The misconfigured WT example is now giving negative overpayments... Sigh.

**Comment by darosior** [comment]
A side effect of this is the significant simulation runtime decrease due to having a lot less coins.

File: model/statemachine.py [hunk 1]
```diff
@@ -1041,72 +1048,72 @@ def cancel_coin_selec_1(self, vault, needed_fee, feerate):
...
-            f" {feerate}"
-        )
-        return select_coins(vb_coins + vm_coins)
+            if len(allocated_coins) == 0 or largest_coin_effective_value <= 0:
+                logging.error(
+                    "     Not enough coin allocated to select for Cancel."
+                    f" {needed_fee}sats still needed after selecting "
+                    f"{selected_coins}."
+                )
+                return selected_coins
```
Comment (darosior) [inline model/statemachine.py:1059]: Should use `select_coins`!

**Comment by JSwambo** [comment]
> We should really ditch all the other coin selection but the one not consolidating.. All the heuristics they use are flawed.

I think they are flawed with the new coin distribution too. I think we should have _some_ heuristic for consolidation though. If not... dust coins can accumulate.

The per-vault reserve dist is topped up as needed, and (I think) those coins will not become "dust" (at our anticipated max fee-rate) based on their definition at time of creation. They _could_ become "dust" if we underestimate the max-feerate, in which case new coins which top-up the reserve dist must be created (with refill & cf transition). "dust" reserve coins just move to the bonus reserve. 

The per-vault bonus reserve is more likely to accumulate _actual dust_ over time, and I think those should be consolidated. Any coin that isn't able to bump the feerate of a cancel at the current feerate (or past 90 days, or similar) should considered as dust.

**Comment by darosior** [comment]
> If not... dust coins can accumulate.

I don't disagree, however in all simulations i ran looking closely at the allocated coins dist the supply of small coins runs out pretty quickly. In practice we have large-ish small coins that get consumed most of the time during normal fee periods.

> The per-vault reserve dist is topped up as needed, and (I think) those coins will not become "dust"

I agree. One aspect that we explicitly did not consider but turns out to be more important than i expected is about too large coins. Considering the current exchange rate of bitcoins compared to when we started this study, fees are higher denominated in $ but lower in BTC: we might have to split too-large coins in the future. Even, in practice, not have a perpetually increasing reserve feerate..

But all that is pretty far-fetched and i am still biased toward "price won't go up forever and CUMMAX is still what makes most sense" :).

> or past 90 days

I think it's what makes the most sense.

**Comment by darosior** [comment]
Ok i forgot to push :sob:. Here you go

**Comment by darosior** [comment]
Now cleaning up the CF coin selections

**Review by JSwambo — APPROVED** [review]
ACK 32db9e5a40e277d373687a2a9c8c59e83e9fb4e4

**Pull request merged**